build: intermittent build failures when patching external sources

this is intermittent. a rebuild usually resolves it.

```
-- zlib download command succeeded.  See also /home/rook/go/src/github.com/rook/rook/external/build/x86_64-linux-gnu/src/zlib-stamp/zlib-download-*.log
[ 21%] No update step for 'aio'
[ 22%] Performing patch step for 'aio'
-- External: Building for aarch64-linux-gnu
-- External: Build Type: RelWithDebInfo
-- External: CFLAGS:  -g -O2 -DNDEBUG -fPIC
-- External: CXXFLAGS:  -g -O2 -DNDEBUG -fPIC
-- External: LDFLAGS:  
-- External: Building Zlib 1.2.11
can't find file to patch at input line 3
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|--- src/Makefile	2016-11-30 09:59:12.889526296 -0800
|+++ src/Makefile	2016-11-30 10:00:46.812341855 -0800
--------------------------
File to patch: 
Skip this patch? [y] 
Skipping patch.
3 out of 3 hunks ignored
```

another instance:

```

[ 41%] No update step for 'gperftools'
-- Configuring done
CMake Error at /home/rook/go/src/github.com/rook/rook/.work/build/x86_64-linux-gnu/src/jemalloc-stamp/jemalloc-configure-RelWithDebInfo.cmake:16 (message):
  Command failed: No such file or directory

   './configure' 'CC=/usr/lib/ccache/x86_64-linux-gnu-gcc' 'CXX=/usr/lib/ccache/x86_64-linux-gnu-g++' 'AR=/usr/bin/x86_64-linux-gnu-ar' 'RANLIB=/usr/bin/x86_64-linux-gnu-ranlib' 'EXTRA_CFLAGS= -g -O2 -DNDEBUG -fPIC' '--prefix=/home/rook/go/src/github.com/rook/rook/.work/build/x86_64-linux-gnu' '--host=x86_64-linux-gnu'

  See also

    /home/rook/go/src/github.com/rook/rook/.work/build/x86_64-linux-gnu/src/jemalloc-stamp/jemalloc-configure-*.log


packages/jemalloc/CMakeFiles/jemalloc.dir/build.make:106: recipe for target 'src/jemalloc-stamp/jemalloc-configure' failed
make[7]: *** [src/jemalloc-stamp/jemalloc-configure] Error 1
```

and:
```
-- zstd configure command succeeded.  See also /home/rook/go/src/github.com/rook/rook/.work/build/aarch64-linux-gnu/src/zstd-stamp/zstd-configure-*.log
CMake Error at /home/rook/go/src/github.com/rook/rook/.work/build/aarch64-linux-gnu/src/snappy-stamp/snappy-configure-RelWithDebInfo.cmake:16 (message):
  Command failed: No such file or directory

   './configure' 'CC=/usr/lib/ccache/aarch64-linux-gnu-gcc' 'CXX=/usr/lib/ccache/aarch64-linux-gnu-g++' 'AR=/usr/bin/aarch64-linux-gnu-ar' 'RANLIB=/usr/bin/aarch64-linux-gnu-ranlib' 'CXXFLAGS= -g -O2 -DNDEBUG -fPIC' '--prefix=/home/rook/go/src/github.com/rook/rook/.work/build/aarch64-linux-gnu' '--host=aarch64-linux-gnu'

  See also

    /home/rook/go/src/github.com/rook/rook/.work/build/aarch64-linux-gnu/src/snappy-stamp/snappy-configure-*.log


packages/snappy/CMakeFiles/snappy.dir/build.make:106: recipe for target 'src/snappy-stamp/snappy-configure' failed
make[7]: *** [src/snappy-stamp/snappy-configure] Error 1
CMakeFiles/Makefile2:195: recipe for target 'packages/snappy/CMakeFiles/snappy.dir/all' failed
make[6]: *** [packages/snappy/CMakeFiles/snappy.dir/all] Error 2
make[6]: *** Waiting for unfinished jobs....
[ 21%] Performing build step for 'aio'
github.com/rook/rook/vendor/github.com/spf13/cobra
[ 22%] Performing build step for 'zstd'
CMake Error at /home/rook/go/src/github.com/rook/rook/.work/build/aarch64-linux-gnu/src/zstd-stamp/zstd-build-RelWithDebInfo.cmake:16 (message):
  Command failed: 2

   'make' '-C' '/home/rook/go/src/github.com/rook/rook/.work/build/aarch64-linux-gnu/src/zstd/lib' 'CC=/usr/lib/ccache/aarch64-linux-gnu-gcc' 'CXX=/usr/lib/ccache/aarch64-linux-gnu-g++' 'AR=/usr/bin/aarch64-linux-gnu-ar' 'RANLIB=/usr/bin/aarch64-linux-gnu-ranlib' 'CFLAGS= -g -O2 -DNDEBUG -fPIC'

  See also

    /home/rook/go/src/github.com/rook/rook/.work/build/aarch64-linux-gnu/src/zstd-stamp/zstd-build-*.log


packages/zstd/CMakeFiles/zstd.dir/build.make:111: recipe for target 'src/zstd-stamp/zstd-build' failed
make[7]: *** [src/zstd-stamp/zstd-build] Error 1
```

this might be happening when there are concurrent builds going that share the download dir

I can repro this consistently by running the following:

```
build/run
cd external
make -j32 cross
```

here's a sample of the log files during error:

```
rook@moby:~/go/src/github.com/rook/rook/external$ cat /home/rook/go/src/github.com/rook/rook/external/.work/build/x86_64-linux-gnu/src/zstd-stamp/zstd-download-*.log
CMake Error: Problem with archive_read_next_header(): Truncated input file (needed 18432 bytes, only 0 available)
CMake Error: Problem extracting tar: /home/rook/go/src/github.com/rook/rook/external/.download/v1.1.3.tar.gz
CMake Error at zstd-stamp/extract-zstd.cmake:33 (message):
  error: extract of
  '/home/rook/go/src/github.com/rook/rook/external/.download/v1.1.3.tar.gz'
  failed


CMake Error at zstd-stamp/zstd-download-RelWithDebInfo-impl.cmake:29 (message):
  Command failed (1):

   '/usr/bin/cmake' '-P' '/home/rook/go/src/github.com/rook/rook/external/.work/build/x86_64-linux-gnu/src/zstd-stamp/extract-zstd.cmake'


-- Downloading...
   dst='/home/rook/go/src/github.com/rook/rook/external/.download/v1.1.3.tar.gz'
   timeout='none'
-- Using src='https://github.com/facebook/zstd/archive/v1.1.3.tar.gz'
-- [download 0% complete]
-- [download 1% complete]
-- [download 2% complete]
-- [download 3% complete]
-- [download 4% complete]
-- [download 5% complete]
-- [download 6% complete]
-- [download 7% complete]
-- [download 8% complete]
-- [download 9% complete]
-- [download 10% complete]
-- [download 11% complete]
-- [download 12% complete]
-- [download 13% complete]
-- [download 14% complete]
-- [download 15% complete]
-- [download 16% complete]
-- [download 17% complete]
-- [download 18% complete]
-- [download 19% complete]
-- [download 20% complete]
-- [download 21% complete]
-- [download 22% complete]
-- [download 23% complete]
-- [download 24% complete]
-- [download 25% complete]
-- [download 26% complete]
-- [download 27% complete]
-- [download 28% complete]
-- [download 29% complete]
-- [download 30% complete]
-- [download 31% complete]
-- [download 32% complete]
-- [download 33% complete]
-- [download 34% complete]
-- [download 35% complete]
-- [download 37% complete]
-- [download 39% complete]
-- [download 41% complete]
-- [download 43% complete]
-- [download 45% complete]
-- [download 47% complete]
-- [download 49% complete]
-- [download 51% complete]
-- [download 53% complete]
-- [download 54% complete]
-- [download 55% complete]
-- [download 56% complete]
-- [download 57% complete]
-- [download 58% complete]
-- [download 59% complete]
-- [download 60% complete]
-- [download 61% complete]
-- [download 62% complete]
-- [download 63% complete]
-- [download 64% complete]
-- [download 65% complete]
-- [download 66% complete]
-- [download 67% complete]
-- [download 68% complete]
-- [download 69% complete]
-- [download 70% complete]
-- [download 71% complete]
-- [download 72% complete]
-- [download 73% complete]
-- [download 74% complete]
-- [download 75% complete]
-- [download 76% complete]
-- [download 77% complete]
-- [download 78% complete]
-- [download 79% complete]
-- [download 80% complete]
-- [download 81% complete]
-- [download 82% complete]
-- [download 83% complete]
-- [download 84% complete]
-- [download 85% complete]
-- [download 86% complete]
-- [download 87% complete]
-- [download 88% complete]
-- [download 89% complete]
-- [download 90% complete]
-- [download 91% complete]
-- [download 92% complete]
-- [download 93% complete]
-- [download 94% complete]
-- [download 95% complete]
-- [download 96% complete]
-- [download 97% complete]
-- [download 98% complete]
-- [download 99% complete]
-- [download 100% complete]
-- Downloading... done
-- extracting...
     src='/home/rook/go/src/github.com/rook/rook/external/.download/v1.1.3.tar.gz'
     dst='/home/rook/go/src/github.com/rook/rook/external/.work/build/x86_64-linux-gnu/src/zstd'
-- extracting... [tar xfz]
-- extracting... [error clean up]
rook@moby:~/go/src/github.com/rook/rook/external$ 
```

It does look like a race when two threads are attempting to download and extract at the same time.

--- a/external/Makefile
+++ b/external/Makefile
@@ -52,14 +52,18 @@ ifeq ($(CCACHE),1)
 endif
 
 # The working directory where build generated files are stored
-WORKDIR ?= .work
+WORKDIR ?= $(abspath .work)
 BUILDDIR= $(WORKDIR)/build
 
 # A directory to hold files that are downloaded
 # as part of the build process but are immutable and do not
 # vary across build runs. These are can be cleaned with
 # "make distclean"
-DOWNLOADDIR ?= .download
+DOWNLOADDIR ?= $(abspath .download)
+
+# workaround races with cmake by copying the download dirs
+# to a private directory
+PRIVATE_DOWNLOADDIR = $(WORKDIR)/download
 
 # The host platform
 HOST = $(shell gcc -dumpmachine)
@@ -69,14 +73,16 @@ HOST = $(shell gcc -dumpmachine)
 
 define build-platform
 build.$(1):
+	@mkdir -p $(PRIVATE_DOWNLOADDIR)/$(1)
 	@mkdir -p $(DOWNLOADDIR)
+	@cp -a $(DOWNLOADDIR)/. $(PRIVATE_DOWNLOADDIR)/$(1)
 	@cmake \
 		-H$(CURDIR) \
 		-B$(BUILDDIR)/$(1) \
 		-DALLOCATOR=$(ALLOCATOR) \
 		-DCEPH_BRANCH=$(CEPH_BRANCH) \
 		-DEXTERNAL_ALWAYS_BUILD=$(ALWAYS_BUILD) \
-		-DEXTERNAL_DOWNLOAD_DIR=$(DOWNLOADDIR)/$(1) \
+		-DEXTERNAL_DOWNLOAD_DIR=$(PRIVATE_DOWNLOADDIR)/$(1) \
 		-DCMAKE_TOOLCHAIN_FILE=$(CURDIR)/toolchain/$(1).cmake \
 		-DEXTERNAL_LOGGING=$(EXTERNAL_LOGGING) \
 		-DEXTERNAL_CCACHE=$(EXTERNAL_CCACHE) \
@@ -84,6 +90,7 @@ build.$(1):
 		-DCMAKE_POSITION_INDEPENDENT_CODE=ON
 
 	@+$(MAKE) -C $(BUILDDIR)/$(1)
+	@cp -na $(PRIVATE_DOWNLOADDIR)/$(1)/. $(DOWNLOADDIR)
 endef
 
 $(foreach p,$(PLATFORMS),$(eval $(call build-platform,$(p))))

--- a/external/packages/aio/CMakeLists.txt
+++ b/external/packages/aio/CMakeLists.txt
@@ -15,6 +15,7 @@
 # See https://git.fedorahosted.org/cgit/libaio.git/
 set(Aio_VERSION 0.3.110)
 set(Aio_URL http://ftp.us.debian.org/debian/pool/main/liba/libaio/libaio_${Aio_VERSION}.orig.tar.gz)
+set(Aio_SHA256 e019028e631725729376250e32b473012f7cb68e1f7275bfc1bbcdd0f8745f7e)
 
 message(STATUS "External: Building libaio ${Aio_VERSION}")
 
@@ -22,11 +23,11 @@ message(STATUS "External: Building libaio ${Aio_VERSION}")
 # Build
 #
 
-
 ExternalProject_Add(aio
   PREFIX ${EXTERNAL_ROOT}
 
   URL ${Aio_URL}
+  URL_HASH SHA256=${Aio_SHA256}
 
   DOWNLOAD_DIR ${EXTERNAL_DOWNLOAD_DIR}
   BUILD_IN_SOURCE 1

--- a/external/packages/gperftools/CMakeLists.txt
+++ b/external/packages/gperftools/CMakeLists.txt
@@ -15,6 +15,7 @@
 # See https://github.com/gperftools/gperftools
 set(Gperftools_VERSION 2.5)
 set(Gperftools_URL https://github.com/gperftools/gperftools/releases/download/gperftools-${Gperftools_VERSION}/gperftools-${Gperftools_VERSION}.tar.gz)
+set(Gperftools_SHA256 6fa2748f1acdf44d750253e160cf6e2e72571329b42e563b455bde09e9e85173)
 
 message(STATUS "External: Building gperftools ${Gperftools_VERSION}")
 
@@ -31,6 +32,7 @@ ExternalProject_Add(gperftools
   PREFIX ${EXTERNAL_ROOT}
 
   URL ${Gperftools_URL}
+  URL_HASH SHA256=${Gperftools_SHA256}
 
   DOWNLOAD_DIR ${EXTERNAL_DOWNLOAD_DIR}
   BUILD_IN_SOURCE 1

--- a/external/packages/jemalloc/CMakeLists.txt
+++ b/external/packages/jemalloc/CMakeLists.txt
@@ -15,6 +15,7 @@
 # See https://github.com/google/snappy
 set(Jemalloc_VERSION 4.3.1)
 set(Jemalloc_URL https://github.com/jemalloc/jemalloc/releases/download/${Jemalloc_VERSION}/jemalloc-${Jemalloc_VERSION}.tar.bz2)
+set(Jemalloc_SHA256 f7bb183ad8056941791e0f075b802e8ff10bd6e2d904e682f87c8f6a510c278b)
 
 message(STATUS "External: Building jemalloc ${Jemalloc_VERSION}")
 
@@ -32,6 +33,7 @@ ExternalProject_Add(jemalloc
   PREFIX ${EXTERNAL_ROOT}
 
   URL ${Jemalloc_URL}
+  URL_HASH SHA256=${Jemalloc_SHA256}
 
   DOWNLOAD_DIR ${EXTERNAL_DOWNLOAD_DIR}
   BUILD_IN_SOURCE 1

--- a/external/packages/rocksdb/CMakeLists.txt
+++ b/external/packages/rocksdb/CMakeLists.txt
@@ -14,7 +14,7 @@
 
 set(Rocksdb_HASH a0deec960f3a8190831c673e5ba998fe6fb7ea90)
 set(Rocksdb_URL https://github.com/facebook/rocksdb/archive/${Rocksdb_HASH}.zip)
-
+set(Rocksdb_SHA256 c6c0226ef75fa73ead5b6bd259599cd4653fcb89e1d4aa6f22432beae9fb08b5)
 
 message(STATUS "External: Building rocksdb ${Rocksdb_HASH}")
 
@@ -31,6 +31,7 @@ ExternalProject_Add(rocksdb
   PREFIX ${EXTERNAL_ROOT}
 
   URL ${Rocksdb_URL}
+  URL_HASH SHA256=${Rocksdb_SHA256}
 
   DOWNLOAD_DIR ${EXTERNAL_DOWNLOAD_DIR}
   DOWNLOAD_NAME rocksdb-${Rocksdb_HASH}.zip

--- a/external/packages/snappy/CMakeLists.txt
+++ b/external/packages/snappy/CMakeLists.txt
@@ -15,6 +15,7 @@
 # See https://github.com/google/snappy
 set(Snappy_VERSION 1.1.4)
 set(Snappy_URL https://github.com/google/snappy/releases/download/${Snappy_VERSION}/snappy-${Snappy_VERSION}.tar.gz)
+set(Snappy_SHA256 134bfe122fd25599bb807bb8130e7ba6d9bdb851e0b16efcb83ac4f5d0b70057)
 
 message(STATUS "External: Building snappy ${Snappy_VERSION}")
 
@@ -31,6 +32,7 @@ ExternalProject_Add(snappy
   PREFIX ${EXTERNAL_ROOT}
 
   URL ${Snappy_URL}
+  URL_HASH SHA256=${Snappy_SHA256}
 
   DOWNLOAD_DIR ${EXTERNAL_DOWNLOAD_DIR}
   BUILD_IN_SOURCE 1

--- a/external/packages/zstd/CMakeLists.txt
+++ b/external/packages/zstd/CMakeLists.txt
@@ -15,6 +15,7 @@
 # See https://github.com/facebook/zstd
 set(Zstd_VERSION 1.1.3)
 set(Zstd_URL https://github.com/facebook/zstd/archive/v${Zstd_VERSION}.tar.gz)
+set(Zstd_SHA256 106c532ae840a6ee4aee5258f04f3acab7b3e09b9e9584ebe94e4fbfd899af0a)
 
 message(STATUS "External: Building zstd ${Zstd_VERSION}")
 
@@ -26,6 +27,7 @@ ExternalProject_Add(zstd
   PREFIX ${EXTERNAL_ROOT}
 
   URL ${Zstd_URL}
+  URL_HASH SHA256=${Zstd_SHA256}
 
   DOWNLOAD_DIR ${EXTERNAL_DOWNLOAD_DIR}
   BUILD_IN_SOURCE 1

